projects/pgai/pgai/vectorizer/processor.py

```diff
@@ -266,4 +280,152 @@ async def run(self) -> Exception | None:
+            else:
+                # sleep for the poll interval or until the shutdown event is set
+                try:
+                    await asyncio.wait_for(
+                        shutdown_event.wait(), timeout=self.poll_interval
+                    )
+                    # shutdown event was set, the loop should exit
+                    logger.info("got a graceful shutdown request")
+                except asyncio.TimeoutError:
+                    # timeout means the sleep completed
+                    pass
+
+        if worker_tracking is not None:
+            await worker_tracking.force_last_heartbeat_and_stop()
+        logger.info("exiting processor.run()")
+        return None
+
+
+class ProcProcessor(Processor):
+    wait_timeout = 300
+
+    @dataclass
+    class StartupMsg:
+        pass
+
+    @dataclass
+    class ShutdownMsg:
+        exception: Exception | None
+
+    QueueType = mp.Queue
+    MessageType = StartupMsg | ShutdownMsg
+
+    def _create_graceful_shutdown_event_using_signals(self):
+        shutdown_event = asyncio.Event()
+        loop = asyncio.get_running_loop()
+
+        def _shutdown_signal_handler(shutdown_event: asyncio.Event, sig: int):
+            """Signal handler that sets the asyncio event"""
+            shutdown_event.set()
+
+            # SIGINT is forcibly shutdown, SIGTERM is graceful shutdown
+            if sig == signal.SIGINT:
+                signame = signal.Signals(signal.SIGINT).name
+                logger.info(f"received {signame} in ProcProcessor, exiting")
+                sys.exit(1)
+
+        for sig in (signal.SIGINT, signal.SIGTERM):
+            loop.add_signal_handler(
+                sig, functools.partial(_shutdown_signal_handler, shutdown_event, sig)
+            )
+        return shutdown_event
+
+    async def _run_inside_new_process(
+        self, result_queue: "ProcProcessor.QueueType[ProcProcessor.MessageType]"
+    ) -> Exception | None:
+        shutdown_event = self._create_graceful_shutdown_event_using_signals()
+        result_queue.put(ProcProcessor.StartupMsg())
+        return await self.run(shutdown_event)
+
+    @staticmethod
+    def _run_processor_inside_new_process(
+        processor: "ProcProcessor",
+        result_queue: "ProcProcessor.QueueType[ProcProcessor.MessageType]",
+    ) -> Exception | None:
+        import asyncio
+
+        result: Exception | None = asyncio.run(
+            processor._run_inside_new_process(result_queue)
+        )
+        result_queue.put(ProcProcessor.ShutdownMsg(result))
+        return result
+
+    def run_in_new_process(self):
+        # todo: monitor for process termination
+        self.result_queue: ProcProcessor.QueueType[ProcProcessor.MessageType] = (
+            mp.Queue()
+        )
+        self.process = mp.Process(
+            target=self._run_processor_inside_new_process,
+            args=(self, self.result_queue),
+        )
+        self.process.start()
```

Comment:
Is that the correct way to let it run alongside the fastapi process? I would have thought we try to hook our existing worker into the fastapi eventloop instead of starting another process?
I feel like if you have two processes you probably should have two containers for it?

Reply:
TBH I would be slightly concerned with putting the vectorizer in fastapi's event loop. While we are using async, I don't think that we can know that we're (or our dependencies) not accidentally doing blocking calls somewhere.

I've looked for other examples of doing "background work" in FastAPI and found two:
- APScheduler: https://pypi.org/project/APScheduler/
- FastAPI Utils "Repeated Tasks": https://fastapi-utils.davidmontague.xyz/user-guide/repeated-tasks/

APScheduler offers both asyncio integration, and a thread-based background worker. FastAPI Utils also does both, from what I can tell, it runs async functions in the event loop, and synchronous functions in a thread pool.